### PR TITLE
fix: handle deletions more consistently in lexicon and rules

### PR DESCRIPTION
**Pull request by dhdaines**
This fixes a couple of things, actually:

- `substring_alignments` wasn't entirely robust in its handling of deletions (i.e. edges with `None` in second position)
- BUT ... these edges actually never occur with rule-based mappings because `None` is systematically (but somewhat arbitrarily) removed in every case unless there is no output to which a deletion could actually be mapped
- BUT ... it wasn't being systematically removed in some cases (dealing with concatenation) so this fixes that
- AND ... there was a bug in the code that did that which could result in an IndexError so this fixes that too
- and now the lexicon-based mappings handle deletions in the same way as the rules
- and the type annotations indicate where `None` can occur (this could still be improved somewhat)

### CODE REVIEW

File: g2p/transducer/__init__.py [hunk 1]
```diff
@@ -264,7 +266,10 @@ def find_monotonic_segments(alignments):
                 else:
                     assert oedge[0] is not None
                     iend = max(iend, oedge[0])
-                    if iedge[1] is not None:
+                    # Replace None with not-None
+                    if oend is None:
+                        oend = iedge[1]
```
Comment (joanise) [inline g2p/transducer/__init__.py:271]: I don't fully understand the logic, and this specific case is not covered by unit testing (maybe add a test to cover it?) but is it using the right non-None thing? Are you mapping to the nearest thing you can?

File: g2p/transducer/__init__.py [hunk 2]
```diff
@@ -751,29 +756,37 @@ def apply_lexicon(self, to_convert: str):
...
+                (n_inputs, outtxt) = alignment[idx : idx + 2]
                 for i in range(n_inputs):
                     for j in range(len(outtxt)):
                         edges.append((in_pos + i, out_pos + j))
-                    if len(outtxt) == 0:  # Deletions
-                        edges.append((in_pos + i, None))
+                    if len(outtxt) == 0:
+                        # Attach deletions to the next input and the
+                        # previous output (fixed below if it does not
+                        # exist)
```
Comment (joanise) [inline g2p/transducer/__init__.py:771]: This comment is confusing. By "next input", do you mean the current one being processed? Do you just mean "Attach deletions to the previous output"?

File: g2p/transducer/__init__.py (hunk 1 above)
Comment (dhdaines) [reply g2p/transducer/__init__.py:271]: Oh - I should make sure this gets covered.  We have to always ensure that no `None` escapes into the computed alignments - since a `None` means a deletion, it would be replaced with the next available index.  In practice this may not happen but I think I can still provoke it in a test case, so I will do that.

File: g2p/transducer/__init__.py (hunk 2 above)
Comment (dhdaines) [reply g2p/transducer/__init__.py:771]: Yes, the comment here is actually wrong.  Obviously the input for a deletion is just the deleted input.  But also it isn't actually attaching them to the previous output but rather the *next* one here.  The choice of previous or next is entirely arbitrary but probably it should use the previous one to be consistent with what rules do (even though in the current english lexicon this gives us considerably weirder alignments).

File: g2p/transducer/__init__.py (hunk 1 above)
Comment (dhdaines) [reply g2p/transducer/__init__.py]: this got removed because we now ensure `None`  doesn't make it there upstream from here

File: g2p/mappings/utils.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 GEN_DIR = os.path.join(os.path.dirname(langs.__file__), "generated")
 GEN_CONFIG = os.path.join(GEN_DIR, "config.yaml")
+IntOrNone = TypeVar("IntOrNone", bound=Union[int, None])
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py:24]: Wow, no, it really isn't! I don't get it. I just tried, to see what would happen, and mypy complains all over!

Why does mypy accept an IntOrNone where it expects an int, but not an Optional[int] in the same situation? The very definition of Optional[X] is Union[X, None], which is what you bind IntOrNone to!

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py:24]: It gets more confusion by the second. mypy won't let me assign a Union[int, None] to an int. What magic did you do to get it to let you assign an IntOrNone to an int?

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py:24]: I just read up on what bound means in TypeVar, but I still don't get it.
Mypy even lets me do this:
```
v1: IntOrNone = None
v2: int = v1
```
this clearly assigns None to v2, which should not have been allowed. Is this a mypy bug?

**Review by joanise — APPROVED** [review]
I'm immensely confused about the semantics of IntOrNone (see inline comments) but otherwise this looks good and ready to merge in my opinion.

File: g2p/mappings/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py:24]: OK, now I get it. I've done some more narrow tests in isolation, and I see that it's for function arguments, and the narrower type gets bound are call time based on the caller's actual parameters. That make a lot more sense now.

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py:24]: And when I was testing, I failed to notice that `v1: IntOrNone` produces its own error, complaining the type variable `IntOrNone` itself is unbound, and therefore no, it's not true mypy lets me do `v2: int = v1`, it's just that that line is not where the error is!

[the remark this answers is not in the document]
Comment (dhdaines) [reply g2p/mappings/utils.py:24]: Hi! Yes, it's all very confusing, but basically mypy doesn't consider `List[Tuple[int, int]]` to be a subtype of `List[Tuple[int, Optional[int]]]` for some reason that I don't entirely grasp (since it *does* allow you to assign a `Tuple[int, int]` to a `Tuple[int, Optional[int]]`).  Something about invariant variant covariant mumble mumble mumble.

While we might be able to get around this by using `Sequence` instead of `List`, the advantage of using a type variable here is that if you pass a `List[Tuple[int, int]]` to `compose_indices`, the return type will be `List[Tuple[int, int]]` instead of `List[Tuple[int, Optional[int]]]`, which is what we want in a lot of cases where we *know* there won't be a `None` there.

[the remark this answers is not in the document]
Comment (dhdaines) [reply g2p/mappings/utils.py:24]: I will rename `IntOrNone` to `IntOrOptionalInt` maybe ;-)

[the remark this answers is not in the document]
Comment (dhdaines) [reply g2p/mappings/utils.py:24]: to illustrate this, you can try running mypy on:

```python
from typing import Tuple, List, Sequence, Optional

a: Tuple[int, int] = (1, 2)
b: Tuple[int, Optional[int]] = (1, 2)
c: Tuple[int, Optional[int]] = a
# this would fail
# d: Tuple[int, int] = b


aa: List[Tuple[int, int]] = [a, a]
ab: List[Tuple[int, Optional[int]]] = [a, b]

# Sequence works here, List doesn't
aaaaaa: Sequence[Tuple[int, Optional[int]]] = aa
# this would fail
# aaaaaa: List[Tuple[int, Optional[int]]] = aa
```

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/utils.py]: Thanks for this. Now that I understand what bound TypeVars actually, this makes a lot of sense. I find it really neat how they let you assign the output type to the input that, I can see that being really useful. And it's all still static-checkable since you get that input (and therefore output) type at each place when a function gets called.

I like the name IntOrOptionalInt, it makes the intent clear.
